### ISSUE TITLE
dcm2niix windows

### ISSUE THREAD

**Issue by ayalalazaro** [wasserth/TotalSegmentator#106]
Firstly, congratulations for your awesome job, I have installed and run TotalSegmentator in Ubuntu with no issues but in windows there is a problem when looking for dcm2niix. 
in dicom_io.py:
```
def dcm_to_nifti(input_path, output_path, verbose=False):
  ...
  dcm2niix = config_dir / "dcm2niix"
  # It should be substituted with:
  dcm2niix = config_dir / "dcm2niix.exe"
```
Otherwise it never finds dcm2niix.

**Comment by wasserth** [comment]
Thank you for your kind words and for bringing up the issue with the `dcm2niix` executable on Windows. You are correct that it requires the `.exe` extension when running on Windows. I just updated the code accordingly. Thanks again for bringing this to our attention!

**Comment by ayalalazaro** [comment]
My pleasure!

**Comment by xiaoli26** [comment]
Though I succeeded in the end somehow, I will try this method next time, thanks!

**Comment by KwintenTorfs** [comment]
Great prospect in using this amazing program!

I did find some problems when trying to install the package in Python (pycharm) on a windows PC.
First off, the dcm2niix issue aswell.

But later on also a problem in the dcm_to_nifti function where the os.remove(str(output_path)[:-7] + ".json") throws a filenotfounderror. Any suggestions/ideas on how to solve?

**Comment by wasserth** [comment]
Could you post the complete error message?

**Comment by KwintenTorfs** [comment]
Thanks for the quick response. This is what I'm experiencing

System cannot  find the given path
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm 2022.3\plugins\python\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "C:\Program Files\JetBrains\PyCharm 2022.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2022.3\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\Users\ktorfs5\KU Leuven\Projects Pycharm\segmentationTRIAL\main.py", line 7, in <module>
    seg = totalsegmentator(inp, outp)
  File "C:\Users\ktorfs5\KU Leuven\Projects Pycharm\segmentationTRIAL\venv\lib\site-packages\totalsegmentator\python_api.py", line 193, in totalsegmentator
    seg_img = nnUNet_predict_image(input, output, task_id, model=model, folds=folds,
  File "C:\Users\ktorfs5\KU Leuven\Projects Pycharm\segmentationTRIAL\venv\lib\site-packages\totalsegmentator\nnunet.py", line 170, in nnUNet_predict_image
    dcm_to_nifti(file_in, tmp_dir / "dcm" / "converted_dcm.nii.gz", verbose=verbose)
  File "C:\Users\ktorfs5\KU Leuven\Projects Pycharm\segmentationTRIAL\venv\lib\site-packages\totalsegmentator\dicom_io.py", line 95, in dcm_to_nifti
    os.remove(str(output_path)[:-7] + ".json")
FileNotFoundError: [WinError 2] System cannot find the given file: 'C:\\Users\\ktorfs5\\AppData\\Local\\Temp\\nnunet_tmp_rkzpeemh\\dcm\\converted_dcm.json'

**Comment by ahlaughland** [comment]
Thanks for putting this great project together! I'm looking forward to getting it up an running.

I'm getting the same error as noted by [KwintenTorfs](https://github.com/KwintenTorfs) above using Windows 10:

```
Using 'fast' option: resampling to lower resolution (3mm)
Converting dicom to nifti...
  Downloading dcm2niix...
Traceback (most recent call last):
  File "/code/src/main.py", line 13, in <module>
    main()
  File "/code/src/main.py", line 9, in main
    segmentor.process_segmentation()
  File "/code/src/processing/segmentator.py", line 134, in process_segmentation
    args.force_split, args.output_type, args.quiet, args.verbose, args.test)
  File "/opt/conda/lib/python3.6/site-packages/totalsegmentator/python_api.py", line 208, in totalsegmentator
    quiet=quiet, verbose=verbose, test=test)
  File "/opt/conda/lib/python3.6/site-packages/totalsegmentator/nnunet.py", line 171, in nnUNet_predict_image
    dcm_to_nifti(file_in, tmp_dir / "dcm" / "converted_dcm.nii.gz", verbose=verbose)
  File "/opt/conda/lib/python3.6/site-packages/totalsegmentator/dicom_io.py", line 100, in dcm_to_nifti
    os.remove(str(output_path)[:-7] + ".json")
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/nnunet_tmp_9ll3wdmh/dcm/converted_dcm.json'
```

Let me know if I can provide any further details.

**Comment by wasserth** [comment]
Which version of TotalSegmentator are you using? Can you try again with the latest version?

**Comment by ahlaughland** [comment]
I pulled using the latest ```master``` branch and rebuild the docker image I'm running. I still get the above error.

So you have a better idea of my setup here is my dockerfile:

```
FROM nvcr.io/nvidia/pytorch:20.10-py3

ENV TOTAL_SEGMENTATOR_BRANCH master  

RUN apt-get update \ 
&& apt-get install ffmpeg libsm6 libxext6 -y  \ 
&& apt-get install xvfb -y  \ 
&& pip install --upgrade pip  \ 
&& pip install flask gunicorn python-git

RUN git clone --depth 1 -b ${TOTAL_SEGMENTATOR_BRANCH} --single-branch  https://github.com/wasserth/TotalSegmentator.git /TotalSegmentator \ 
&& pip install /TotalSegmentator \ 
&& python /TotalSegmentator/totalsegmentator/download_pretrained_weights.py

WORKDIR /app

COPY . /code/src/TotalSegmentator
# my custom code
COPY ./src /code/src

# /dicoms and /segmentations are volumes set by the docker run command
CMD ["python3", "/code/src/main.py", "-i", "/dicoms", "-o", "/segmentations", "--fast"]
```

And the contents of ```main.py```:

```
from processing.segmentator import Segmentator

def main():    
    segmentor = Segmentator()
    segmentor.process_segmentation()

if __name__ == "__main__":
    main()
```

When I start ```run_server.sh``` I can start the server.  Then ```docker exec``` into the running container and run ```TotalSegmentator -i ct.nii.gz -o segmentations``` and can see the segmentations in the ```segmentations``` dir. It's not clear to me why running the aforementioned command works vs. what I'm doing - both run in a docker container. 

Thanks for the quick reply!

**Comment by ahlaughland** [comment]
I only get this error when running the segmentator on .dcm files the example_ct.nii.gz works with no errors.

**Comment by ahlaughland** [comment]
@wasserth I was able to generate a .nii file using the ```--ml``` flag by making the following changes to https://github.com/wasserth/TotalSegmentator/blob/cef672ba1200a14051c7429f8cc7ecd71ee0c7a4/totalsegmentator/nnunet.py#L173

```
            file_in = file_in.NamedTemporaryFile(delete = False)
            file_in.close()
```

It looks like Windows opens a file with a share flag and future open operations pass this share flag. It's not the case with Python. 

https://stackoverflow.com/questions/15588314/cant-access-temporary-files-created-with-tempfile

Hope this helps.

**Comment by wasserth** [comment]
Thanks for pointing this out. But I am not sure how exactly you changed the code. Can you share your new lines of code around line 173.

**Comment by ahlaughland** [comment]
Here is the surrounding code:

```python
 if img_type == "dicom":
            if not quiet: print("Converting dicom to nifti...")
            (tmp_dir / "dcm").mkdir()  # make subdir otherwise this file would be included by nnUNet_predict
            dcm_to_nifti(file_in, tmp_dir / "dcm" / "converted_dcm.nii.gz", verbose=verbose)
            file_in_dcm = file_in
            file_in = tmp_dir / "dcm" / "converted_dcm.nii.gz"
            
            # ahlaughland added the next two lines:
            file_in = file_in.NamedTemporaryFile(delete = False)
            file_in.close()
            
            # if not multilabel_image:
            #     shutil.copy(file_in, file_out / "input_file.nii.gz")
            if not quiet: print(f"  found image with shape {nib.load(file_in).shape}")

```

**Comment by wasserth** [comment]
When I run this I get the error message 
```
AttributeError: 'PosixPath' object has no attribute 'NamedTemporaryFile'
```
Strange, that you do not get this.

**Comment by ahlaughland** [comment]
Did you run ```pip install /app``` after updating the file?

**Comment by wasserth** [comment]
yes, I did. Maybe this only works on windows? I am running on ubuntu.

**Comment by ahlaughland** [comment]
> yes, I did. Maybe this only works on windows? I am running on ubuntu.

Yes, I bet that's why you get the error -  I am on Windows 10.

**Comment by MMdeB** [comment]
Thank you for making this wonderful project!

I also get the [WinError 2] The system cannot find the file specified error of the Temp file. I run it on windows, and I tried the solution of changing the code in the nnunet.py, but this did not change anything. What do you mean with the `pip install /app` code? I do not know what the /app does. 

Thanks!

**Comment by JamMassey** [comment]
> > yes, I did. Maybe this only works on windows? I am running on ubuntu.
> 
> Yes, I bet that's why you get the error - I am on Windows 10.
> 
> Maybe add:
> 
> `from sys import platform` and check the OS type.

I get ```AttributeError: 'WindowsPath' object has no attribute 'NamedTemporaryFile'```. On Windows 10.

**Comment by wasserth** [comment]
This problem should now be solved in v2.0.0
